src/mlpack/tests/main_tests/kde_test.cpp

```diff
@@ -0,0 +1,408 @@
+  * get from the KDE class without any wrappers. Doesn't require normalization.
+ **/
+BOOST_AUTO_TEST_CASE(KDETriangularBallTreeResultsMain)
+{
+  // Datasets.
+  arma::mat reference = arma::randu(3, 300);
+  arma::mat query = arma::randu(3, 100);
+  arma::vec kdeEstimations, mainEstimations;
+  double kernelBandwidth = 3.0;
+  double relError = 0.06;
+
+  kernel::TriangularKernel kernel(kernelBandwidth);
+  metric::EuclideanDistance metric;
+  KDE<kernel::TriangularKernel,
+      metric::EuclideanDistance,
+      arma::mat,
+      tree::BallTree>
+      kde(relError, 0.0, kernel, KDEMode::DUAL_TREE_MODE, metric);
+  kde.Train(reference);
+  kde.Evaluate(query, kdeEstimations);
+
+  // Main estimations.
+  SetInputParam("reference", reference);
+  SetInputParam("query", query);
+  SetInputParam("kernel", std::string("triangular"));
+  SetInputParam("tree", std::string("ball-tree"));
+  SetInputParam("rel_error", relError);
+  SetInputParam("bandwidth", kernelBandwidth);
+
+  mlpackMain();
+
+  mainEstimations = std::move(CLI::GetParam<arma::vec>("predictions"));
+
+  // Check whether results are equal.
+  for (size_t i = 0; i < query.n_cols; ++i)
+    BOOST_REQUIRE_CLOSE(kdeEstimations[i], mainEstimations[i], relError);
+}
+
+/**
+  * Ensure that the estimations we get for KDEMain, are the same as the ones we
+  * get from the KDE class without any wrappers in the monochromatic case.
+ **/
+BOOST_AUTO_TEST_CASE(KDEMonoResultsMain)
+{
+  // Datasets.
+  arma::mat reference = arma::randu(2, 300);
+  arma::vec kdeEstimations, mainEstimations;
+  double kernelBandwidth = 2.3;
+  double relError = 0.05;
+
+  kernel::EpanechnikovKernel kernel(kernelBandwidth);
+  metric::EuclideanDistance metric;
+  KDE<kernel::EpanechnikovKernel,
+      metric::EuclideanDistance,
+      arma::mat,
+      tree::StandardCoverTree>
+    kde(relError, 0.0, kernel, KDEMode::DUAL_TREE_MODE, metric);
+  kde.Train(reference);
+  // Perform monochromatic KDE.
+  kde.Evaluate(kdeEstimations);
+  // Normalize.
+  kdeEstimations /= kernel.Normalizer(reference.n_rows);
+
+  // Main estimations.
+  SetInputParam("reference", reference);
+  SetInputParam("kernel", std::string("epanechnikov"));
+  SetInputParam("tree", std::string("cover-tree"));
+  SetInputParam("rel_error", relError);
+  SetInputParam("bandwidth", kernelBandwidth);
+
+  mlpackMain();
+
+  mainEstimations = std::move(CLI::GetParam<arma::vec>("predictions"));
+
+  // Check whether results are equal.
+  for (size_t i = 0; i < reference.n_cols; ++i)
+    BOOST_REQUIRE_CLOSE(kdeEstimations[i], mainEstimations[i], relError);
+}
+
+/**
+  * Ensuring that absence of input data is checked.
+ **/
+BOOST_AUTO_TEST_CASE(KDENoInputData)
+{
+  // No input data is not provided. Should throw a runtime error.
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+  * Check that there're as many densities in the result as query points.
```

Comment:
I think `the are` is the correct way to say it.

Reply:
Solved in 20230ae14408ae5655ad8f02be2d0a26b3b612b7.